woocommerce/woocommerce
# Fix wrong type passed to `add_meta_boxes` action

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:
This PR fixes an issue reported in #41579 where a `TypeError` is thrown in order edit pages because a wrong type is passed to `add_meta_boxes` and `add_meta_boxes_{screen}` actions in the following file:
[src/Internal/Admin/Orders/Edit.php (lines 165 and 175)](https://github.com/woocommerce/woocommerce/blob/35ba49a1f29bad8a1685ae4a179c7273eee3f298/plugins/woocommerce/src/Internal/Admin/Orders/Edit.php#L165-L175)

The TypeError indicates that the passed parameter must be an instance of `WP_Post`, but instance of `Automattic\WooCommerce\Admin\Overrides\Order` given

The issue is reproducible by a code snippet like this:
```PHP
add_action(
	'add_meta_boxes',
	function (string $postType, \WP_Post $post) {
		// Add a metabox
	},
	10,
	2
);
```
It breaks the order edit page with an fatal uncaught TypeError.
The HPOS system stores an order placeholder entry in the `posts` table with the same ID ([source](https://github.com/woocommerce/woocommerce/blob/35ba49a1f29bad8a1685ae4a179c7273eee3f298/docs/high-performance-order-storage/README.md#placeholder-records)), the fix tries to fetch a `WP_Post` object from a given order id and pass it where a `WP_Post` object is expected.






Closes #41579 .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add a test metabox, using this example snippet in your theme's functions.php or within a mu-plugin:
```PHP
add_action(
	'add_meta_boxes',
	function (string $postType, \WP_Post $post) {
		add_meta_box(
			'foo_bar_baz',
			'Foo',
			function (WP_Post $post): void {
				echo 'Test';
			},
			null,
			'advanced',
			'high',
		);
	},
	10,
	2
);
```
2. Submit an order.
3. Go to WooCommerce > Orders and edit the newly submitted order.
4. You should not face a fatal error and the order edit page should correctly load.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
